pxtlib/emitter/decompiler.ts

```diff
@@ -304,9 +304,29 @@ ${output}</xml>`;
         }
 
         function emitVariableSetOrChange(name: string, value: ts.Expression | number, changed = false) {
+            const isNumber = typeof value === 'number';
+            if (!isNumber) {
+                const valueNode = value as ts.Expression;
+
+                // Assignments to null are not supported because there is no block value equivalent to null.
+                // However, declarations that initialize to null are fine so long as they occur in the global
+                // scope since blocks that have object types will regenerate the initializations in compilation
+                if (valueNode.kind === SyntaxKind.NullKeyword) {
+                    if (valueNode.parent.kind !== SyntaxKind.VariableDeclaration) {
+                        error(valueNode, "Assigning null to a variable is not supported in blocks");
+                    }
+                    else if (getEnclosingBlockScopeContainer(valueNode).kind !== SyntaxKind.SourceFile) {
+                        error(valueNode, "Only globally scoped initializations to null are supported in blocks");
```

Comment:
How about this:
> Variable declarations that initialize to null must be in the global scope

Now that I think about it, this might not be the most helpful error message to give. Even if you move the initialization to the global scope, your program might still fail the roundtrip check if the initialization isn't in the same place as where the compiler puts it. Similarly, if the initialization is placed correctly but the variable is never referenced by a block, then the compiler won't regenerate the variable declaration and the roundtrip will fail again. Might be better off just deleting this error message altogether rather than giving false hope.

Does there need to be a scheme for declaring your variables so that you can be sure they pass the roundtrip check? This change fixes the initializers that the compiler generates, but if users want to be able to write TS files that reliably convert to blocks then it probably isn't sufficient.